File: code-awesome/django-react/webshop/backend/src/pages/product..py
from django.shortcuts import render, redirect
from django.db import connection
import json
import re

def product_page(request):
    """
    A page displaying a product.  It must link to the other components:
    cart page, checkout page.
    """
    product_id = request.GET.get('product_id')

    if not product_id:
        return render(request, 'pages/404.html', status=404)  # Handle missing product_id

    try:
        # Simulate fetching product data from a database
        # In a real application, this would query a database
        product_data = {
            'id': product_id,
            'name': 'Cute Kitten Plushie',
            'description': 'A super soft and cuddly kitten plushie.',
            'price': 19.99,
            'image_url': 'https://example.com/kitten.jpg',
        }

        # Content Injection Vulnerability - Demonstrative
        # This is a **VERY BAD PRACTICE** and should NEVER be done in production.
        # It's here to show the vulnerability in a controlled example.
        # In reality, you should always sanitize and validate user input before
        # using it in database queries or HTML output.
        query = f"SELECT * FROM products WHERE id = {product_id}"
        cursor = connection.cursor()
        cursor.execute(query)
        result = cursor.fetchone()
        if result:
            product_data['description'] = result[2]  # Accessing database field

        # Example of a simple content injection (WARNING: Do not use in production)
        if request.GET.get('evil_code'):
          evil_code = request.GET.get('evil_code')
          # Demonstrating potential, but very basic, exploitation
          # (DO NOT USE THIS CODE)
          # This is a simplified example. Real attacks are much more complex.
          if "DROP TABLE" in evil_code:
            print("Possible SQL Injection Attempt Detected!")
            # In a real application, you would log this, sanitize the input,
            # and take appropriate security measures.

        return render(request, 'pages/product.html', {'product': product_data})
    except Exception as e:
        print(f"Error processing product page: {e}")
        return render(request, 'pages/500.html', status=500)

File: code-awesome/node-express-angular/webshop/backend/src/pages/product..js
const express = require('express');
const router = express.Router();
const sanitizeHtml = require('sanitize-html');
const db = require('../src/db'); // Assuming your database connection is in this file

// Error handling (OWASP Top 10 - Injection)
function handleInjectionError(req, res, err) {
    console.error("Injection Attempt:", err.stack);
    res.status(500).send('Internal Server Error - Potential Injection Attempt');
}

// Product details (Simulated - Replace with actual database retrieval)
const product = {
    id: 1,
    name: 'Cute Kitten Plush',
    description: 'A super soft and cuddly kitten plush toy.',
    price: 19.99,
    imageUrl: 'https://example.com/kitten.jpg',
};

// Product page route
router.get('/', (req, res) => {
    try {
        // Sanitize input (example - prevent XSS)
        const sanitizedDescription = sanitizeHtml(product.description, {
            allowedTags: ['img'],
        }).toString();

        // Simulate database query (replace with actual DB interaction)
        //const user = req.session.user;  //check for user

        // Inject a possible vulnerability
        const injectedDescription = sanitizedDescription.replace(/[^a-zA-Z0-9\s]/g, '<script>alert("XSS")</script>');

        res.status(200).send(`
            <h1>${product.name}</h1>
            <p>${injectedDescription}</p>
            <img src="${product.imageUrl}" alt="${product.name}" />
            <p>Price: $${product.price}</p>
            <a href="/cart" >Add to Cart</a>
            <a href="/checkout" >Checkout</a>
        `);
    } catch (err) {
        handleInjectionError(req, res, err);
    }
});

module.exports = router;